implementations/rust/ockam/ockam_identity/src/secure_channel/encryptor_worker.rs

```diff
@@ -91,33 +141,154 @@ impl EncryptorWorker {
+                .await
+            {
+                Ok(credential) => credential,
+                Err(err) => {
+                    info!(
+                        "Credentials refresh failed for {} and is rescheduled in {} seconds",
+                        self.addresses.encryptor,
+                        self.min_credential_refresh_interval.as_secs()
+                    );
+                    // Will schedule a refresh in self.min_credential_refresh_interval
+                    self.schedule_credentials_refresh(ctx, true).await?;
+                    return Err(err);
+                }
             }
+        } else {
+            return Err(IdentityError::NoCredentialsRetriever.into());
+        };
+
+        let versioned_data = credential.credential.get_versioned_data()?;
+        let data = CredentialData::get_data(&versioned_data)?;
+        self.min_credential_expiration = Some(data.expires_at);
+
+        let msg = RefreshCredentialsMessage {
+            change_history,
+            credentials: vec![credential],
         };
+        let msg = SecureChannelMessage::RefreshCredentials(msg);
+
+        let msg = self.encrypt(ctx, msg).await?;
+
+        info!(
+            "Sending credentials refresh for {}",
+            self.addresses.encryptor
+        );
 
         // Send the message to the decryptor on the other side
         ctx.send_from_address(
             self.remote_route.clone(),
-            encrypted_payload,
+            msg,
             self.addresses.encryptor.clone(),
         )
         .await?;
 
+        self.schedule_credentials_refresh(ctx, false).await?;
+
+        Ok(())
+    }
+
+    async fn send_close_channel(&mut self, ctx: &Context) -> Result<()> {
+        let msg = SecureChannelMessage::Close;
+
+        // Encrypt the message
+        let msg = self.encrypt(ctx, msg).await?;
+
+        // Send the message to the decryptor on the other side
+        ctx.send_from_address(
+            self.remote_route.clone(),
+            msg,
+            self.addresses.encryptor.clone(),
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    /// Schedule a DelayedEvent that will at specific point in time put a message
+    /// into EncryptorWorker's own internal mailbox which it will use as a trigger to get a new
+    /// credential and present it to the other side.
+    async fn schedule_credentials_refresh(&mut self, ctx: &Context, is_retry: bool) -> Result<()> {
+        let min_credential_expiration =
+            if let Some(min_credential_expiration) = self.min_credential_expiration {
+                min_credential_expiration
+            } else {
+                // Do nothing if there is no expiration
+                return Ok(());
+            };
+
+        // Cancel the old event
+        self.credential_refresh_event = None;
+
+        let now = now()?;
+
+        let duration = if min_credential_expiration < now + self.refresh_credential_time_gap {
```

Comment:
Why not simply
```rust
let duration = max(min_credential_expiration - self.refresh_credential_time_gap - now, Duration::from_secs(0));
```

Reply:
Isn't that gonna panic if  self.refresh_credential_time_gap + now > min_credential_expiration?

Reply:
In that case we take a duration of `0`. What would panic?

Reply:
@etorreborre https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6237699b4dafeefc6fe2206869cfdf25

Reply:
I see thanks. It's a bad inherited habit of working with signed integers.